### ISSUE TITLE
Union operator

### ISSUE THREAD

**Issue by max-sixty** [PRQL/prql#656]
This is currently lacking. Possibly it could be as simple as:

```elm
from employees
union
from managers
```

**Comment by zopieux** [comment]
What about other set operators? It seems like `difference` and `intersect` aren't discussed either.

**Comment by Baoqi** [comment]
> 

for the "union", we may need to consider both "union" and "union all".

**Comment by snth** [comment]
Hi,

I have been thinking about this have a proposal which may be a bit of a departure from SQL but which I think could get us some benefits.

SQL allows _table expressions/subqueries_ in a number of places, like on either side of a `JOIN`, `UNION` and in (correlated) subqueries. Leaving the correlated subqueries aside for the moment, I think that part of the reason may be that these were available before the introduction of CTEs. Coming at it from a more modern perspective I question whether we still need that flexibility.

My understanding (mostly from @aljazerzen 's responses to issues rather any look a the compiler code) is that the PRQL compiler works by applying successive transformation functions to frames and producing new frames. By the sounds of it this follows the [Fluent Interface data api pattern](https://en.wikipedia.org/wiki/Fluent_interface) akin to how you would construct a pipeline in Pandas or Ibis. 

I think this is quite a good choice and propose that we stick to this fluent interface API because I because it keeps the flow simple and linear. I also think that there is no loss of generality because what we're giving up can be achieved with CTEs.

So rather than
```elm
from employees
union
from managers
```
we have
```elm
table t2 = (
    from managers
)

from employees
union t2
```
or simply
```elm
from employees
union managers
```

**Comment by snth** [comment]
The way I think of it is that you essentially have one `main` pipeline and any other sub-pipelines that you require you need to factor out into _table expressions/CTEs_. 

I think this is also a place where we could innovate a bit over SQL and introduce some additional ergonomics. For example one nuisance with `UNION`s is that you have to have compatible columns. With some knowledge of the table and column types we could generate `union`s or `intersection`s of column names automatically.

```elm
table t1 = (
   from X
   select a, b
)

table t2 = (
   from X
   select b, c
)

from t1
union all:true columns:intersection t2
```
could be translated to
```sql
(SELECT b FROM t1)
UNION ALL
(SELECT b from t2)
```

and

```elm
table t1 = (
   from X
   select a, b
)

table t2 = (
   from X
   select b, c
)

from t1
union all:false columns:union t2
```
could be translated to
```sql
(SELECT a, b, NULL as c FROM t1)
UNION
(SELECT NULL as a, b, c from t2)
```

Lastly, perhaps the columns parameter could also take an explicit list of column names, like
```elm
from employees_america
union all:false columns:[uuid] employees_europe employees_asia employees_africa
```
gets transpiled to
```sql
(SELECT uuid FROM employees_america)
UNION
(SELECT uuid FROM employees_europe)
UNION
(SELECT uuid FROM employees_asia)
UNION
(SELECT uuid FROM employees_africa)
```
This might be useful for unioning together partitioned tables, especially if the table identifiers could be generated with some sort of glob or regex expression.

Perhaps we should rather apply the YAGNI principle and wait until someone actually requests these extensions. I don't have a use case for them personally myself. We do need to have a **union operator** though with some priority, as it's needed for recursive CTEs (see #407). The recursive CTE requirement might need some refinement of this `union` proposal but I'll address that separately.

**Comment by snth** [comment]
@zopieux I think `intersect` and `difference` can be achieved with `sem join`s and `anti joins`. I'm not sure how widely supported those are but there are correlated subquery patterns with `EXISTS` that can work around this. This will require backend adaptor specific code though so I think we should rather address this in a separate issue in order to keep the `union` implementation simple. I will open an issue for those.

**Comment by max-sixty** [comment]
https://github.com/prql/prql/issues/656#issuecomment-1174713630 has some good suggestions.

The `UNION` operation is fairly different from most operations in SQL, so it makes sense that it feels a bit foreign.

I don't have strong views between 

> ```elm
> from employees
> union
> from managers
> ```

and 

> ```elm
> from employees
> union managers
> ```

...probably with a mild preference for the second.

Ideally we can still have an inline pipeline with 

```elm
from employees
union (
  from managers
  derive x = foo
)
```

---

I think we can have a single positional parameter for `all` / `distinct` / `intersect` (which means a bare `union` won't work).

Does that seem like a reasonable first implementation? (Without having strong views on the more advanced suggestions in the later comments)

**Comment by aljazerzen** [comment]
Just realizing that `union` would be a function that can take two tables as arguments, which means that it can also be called as:
```elm
union (
  from employees
) (
  from managers 
  derive x = foo
)
```
or more concise:
```elm
union (from employees) (from managers | derive x = foo)
```

**Comment by max-sixty** [comment]
Do you think it should be more a function that takes two parameters[^1] — or like a transform like `join` is, connecting the pipeline above?

[^1]: or a list

**Comment by aljazerzen** [comment]
I'm saying that these two things are the same :smile: !

```elm
from employees
union (from managers | derive x = foo)
```
... is equivalent to:
```elm
union (from managers | derive x = foo) (from employees)
```

**Comment by max-sixty** [comment]
Yes, great point!

Technically, we can't then have a list; we'd need exactly two parameters (or to change how we do parameters).

**Comment by aljazerzen** [comment]
That's true and its a shame. I'm guessing that there would be many uses-cases for accepting a list...

**Comment by max-sixty** [comment]
Yeah, I think we can iterate on that along with https://github.com/prql/prql/issues/648 around generalizing over single items & lists. 

I can imagine some syntheses of these — e.g. the final argument of a function can take multiple items as arguments and they're converted to a list (though not sure if that's a good idea tbc)

**Comment by aljazerzen** [comment]
Closed by #894 

Related work continues on #761
